gel/_internal/_is_overload.py

```diff
@@ -0,0 +1,209 @@
+# SPDX-PackageName: gel-python
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright Gel Data Inc. and the contributors.
+
+"""AST transformation to overload the `is` and `is not` operators.
+
+This module provides a decorator that transform a function's source to
+replace `is` and `is not` operations with calls to `__gel_is__` and
+`__gel_is_not__` methods on the left-hand operand. This allows objects
+to customize the behavior of identity checks, which is particularly
+useful for query builder expressions where `is [not] None` needs to
+trigger a TypeError.
+
+The approach is to re-parse the function's source and replace the
+identity checks with calls to methods via an ast.NodeTransformer.
+This avoids complex bytecode inspection or manipulation and preserves
+standard `is [not]` behavior for for objects that do not implement the
+`__gel_is__` or `__gel_is_not__` methods.
+"""
+
+from __future__ import annotations
+from typing import TYPE_CHECKING, Any, ParamSpec, TypeVar
+
+import ast
+import functools
+import inspect
+import textwrap
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+_P = ParamSpec("_P")
+_T = TypeVar("_T")
+
+
+def maybe_overload_is_operator(func: Callable[_P, _T]) -> Callable[_P, _T]:
+    """A decorator that overloads `is` and `is not` operators in a
+    function.
+
+    This decorator rewrites the function's bytecode to replace `is` and
+    `is not` operations with calls to `__gel_is__` and `__gel_is_not__`
+    on the left-hand operand.  This is done on a best effort basis, so
+    if there is any issue accessing function source or tranforming or
+    compiling it, the original function is returned unmodified.
+
+    Args:
+        func: The function to be transformed.
+
+    Returns:
+        The transformed function.
+    """
+    if not callable(func):
+        return func
+
+    if getattr(func, "__gel_is_overloaded__", False):
```

Comment:
can we find a way to cache by code object? file location + hash of `__code__` object bytes?

Reply:
<img width="568" height="362" alt="Image" src="https://github.com/user-attachments/assets/94af05e5-fc83-4431-9e12-74cc24992316" />

Reply:
basically right now you'd be rewriting that ast every time, we want to do that once and then reuse the rewritten code object (need to be careful with closure vars)